test/awa_test_server.ml

```diff
@@ -79,6 +79,15 @@ module Driver = struct
       | None -> poll t
       | Some event -> Ok (t, event)
 
+  let user_auth t userauth success =
+    let* server, reply =
+      if success then
+        Awa.Server.accept_userauth t.server userauth
+      else
+        Awa.Server.reject_userauth t.server userauth
+    in
+    send_msg { t with server } reply
```

Comment:
should this logic be done inside of awa -- i.e. provide a function:

```OCaml
let auth ? success =
      if success then
        Awa.Server.accept_userauth t.server userauth
      else
        Awa.Server.reject_userauth t.server userauth
```

this way we wouldn't repeat that in both the test and the mirage server!?

Reply:
or, maybe to reduce the error conditions, we could provide the Userauth event with "yay" and "nay" (which are `accept_userauth` / `reject_userauth` - which wouldn't need to check the state - i.e. there shouldn't be the error case).

Reply:
(Sorry I pressed the "resolve conversation" button instead of the "comment" button and at the same time lost my comment...)

This is interesting. I think we then need to make sure that the `Server.t` doesn't progress through reading and processing incoming packets?

I think it makes sense to combine `accept_userauth` / `reject_userauth` to a variant that takes a bool

Reply:
> make sure that the `Server.t` doesn't progress

being functional and value-passing, isn't it the case that the effectful layer only has that information -- a mutable Server.t -- and from within the core protocol (server.ml) we're not able to verify that "Server.t has not progressed".

Reply:
Right